### PR TITLE
PriceRegistry accepts multi-gas update

### PR DESCRIPTION
**Pull request by matYang**
Allow a single commit report to contain multiple gas price updates, while still reporting a single gas price for now.

Offchain changes are not backwards-compatible yet, will port these changes into the versioning PR.

### CODE REVIEW

**Comment by andrevmatos** [comment]
Price Updates always go through `CommitStore`, which is unique per lane, i.e. it only goes from a specific source chain to a dest chain;
Q: What's the use case for allowing it to update multiple source gas prices?

**Comment by matYang** [comment]
> Price Updates always go through `CommitStore`, which is unique per lane, i.e. it only goes from a specific source chain to a dest chain; Q: What's the use case for allowing it to update multiple source gas prices?

this gas update model is likely to be changed in the near future as part of scaling initiative

**Comment by RensR** [comment]
> this gas update model is likely to be changed in the near future as part of scaling initiative

There might be a pivot in these plans. Let's discuss if this is still a good short term fix before the medium term fix that would make this obsolete.

imo it still does, as all the future options will use multi-gas updates

File: contracts/src/v0.8/ccip/PriceRegistry.sol [hunk 1]
```diff
@@ -201,12 +201,15 @@ contract PriceRegistry is IPriceRegistry, OwnerIsCreator, TypeAndVersionInterfac
       emit UsdPerTokenUpdated(update.sourceToken, update.usdPerToken, block.timestamp);
     }
 
-    if (priceUpdates.destChainSelector != 0) {
-      s_usdPerUnitGasByDestChainSelector[priceUpdates.destChainSelector] = Internal.TimestampedPackedUint224({
-        value: priceUpdates.usdPerUnitGas,
+    uint256 gasUpdatesLength = priceUpdates.gasPriceUpdates.length;
+
+    for (uint256 i = 0; i < gasUpdatesLength; ++i) {
```
Comment (RensR) [inline contracts/src/v0.8/ccip/PriceRegistry.sol:206]: im missing tests for the multi-gas update scenario

File: contracts/foundry.toml [hunk 2]
```diff
@@ -21,7 +21,7 @@ solc_version = '0.8.19'
 src = 'src/v0.8/ccip'
 test = 'src/v0.8/ccip/test'
 optimizer_runs = 26_000
-deny_warnings = true
```
Comment (RensR) [inline contracts/foundry.toml:24]: nit: just removing this will default it to false

File: contracts/src/v0.8/ccip/libraries/Internal.sol [hunk 3]
```diff
@@ -8,15 +8,19 @@ import {MerkleMultiProof} from "../libraries/MerkleMultiProof.sol";
...
+    GasPriceUpdate[] gasPriceUpdates;
   }
 
   struct TokenPriceUpdate {
     address sourceToken; // Source token
     uint224 usdPerToken; // 1e18 USD per smallest unit of token
   }
 
+  struct GasPriceUpdate {
+    uint64 destChainSelector; // ──╮ Destination chain selector
```
Comment (RensR) [inline contracts/src/v0.8/ccip/libraries/Internal.sol]: This doesn't fit in a slot, probably wrong since the uint224 change

Comment (connorwstein) [reply contracts/src/v0.8/ccip/libraries/Internal.sol]: seems ok, just calldata? Analogous to the token price updates

File: core/services/ocr2/plugins/ccip/commit_inflight.go [hunk 4]
```diff
@@ -68,32 +68,27 @@ func (c *inflightCommitReportsContainer) maxInflightSeqNr() uint64 {
 	return max
 }
 
-// getLatestInflightGasPriceUpdate returns the latest inflight gas price update, and bool flag on if update exists.
-// Note we assume that reports contain either 1 or 0 gas prices.
-// If this assumption is broken, we will need to update this logic.
-func (c *inflightCommitReportsContainer) getLatestInflightGasPriceUpdate() (update, bool) {
+// latestInflightGasPriceUpdates returns a map of the latest gas price updates.
+func (c *inflightCommitReportsContainer) latestInflightGasPriceUpdates() map[uint64]update {
```
Comment (connorwstein) [inline core/services/ocr2/plugins/ccip/commit_inflight.go:72]: 🥇

File: core/services/ocr2/plugins/ccip/commit_reporting_plugin.go [hunk 5]
```diff
@@ -806,6 +807,11 @@ func (r *CommitReportingPlugin) isStaleReport(ctx context.Context, lggr logger.L
 	if !hasGasPriceUpdate && !hasTokenPriceUpdates {
 		return true
 	}
+	// Commit plugin currently only supports 1 gas price per report. If report contains more than 1, reject the report.
+	if len(report.GasPrices) > 1 {
+		lggr.Errorw("Report is stale because it contains more than 1 gas price update", "GasPriceUpdates", report.GasPrices)
```
Comment (connorwstein) [inline core/services/ocr2/plugins/ccip/commit_reporting_plugin.go:812]: I think this check better suited to throwing an error in encodeCommitReportV1_2_0? By the time the report makes it to isStale check it cannot be completely invalid like this

File: core/services/ocr2/plugins/ccip/internal/ccipdata/commit_store_reader.go [hunk 6]
```diff
@@ -97,39 +98,34 @@ func NewCommitStoreReader(lggr logger.Logger, address common.Address, ec client.
...
+		commitStoreABI := abihelpers.MustParseABI(commit_store_1_0_0.CommitStoreABI)
 		return func(report []byte) (*txmgr.TxMeta, error) {
 			commitReport, err := decodeCommitReportV1_0_0(abihelpers.MustGetEventInputs(ReportAccepted, commitStoreABI), report)
 			if err != nil {
 				return nil, err
 			}
 			return commitReportToEthTxMeta(commitReport)
 		}, nil
-		// TODO: 1.2 will split
+	case v1_2_0:
```
Comment (connorwstein) [inline core/services/ocr2/plugins/ccip/internal/ccipdata/commit_store_reader.go:120]: 🥇

File: core/services/ocr2/plugins/ccip/internal/ccipdata/commit_store_v1_2_0.go [hunk 7]
```diff
@@ -1,22 +1,161 @@
...
 	ccipconfig "github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/ccip/config"
+	"github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/ccip/internal/logpollerutil"
 	"github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/ccip/prices"
+	"github.com/smartcontractkit/chainlink/v2/core/services/pg"
 	"github.com/smartcontractkit/chainlink/v2/core/store/models"
 )
 
+var _ CommitStoreReader = &CommitStoreV1_2_0{}
+
+type CommitStoreV1_2_0 struct {
```
Comment (connorwstein) [inline core/services/ocr2/plugins/ccip/internal/ccipdata/commit_store_v1_2_0.go:30]: We should be able to embed *CommitStoreV1_0_0 so all the portions of the ABI that didn't change can be re-used (eg IsBlessed, IsDown, GetLatestPriceEpochAndRound..)

**Review by connorwstein — APPROVED** [review]

**Comment by andrevmatos** [comment]
Just a note to self: this also will need some heavy adaptations on Atlas.

File: core/services/ocr2/plugins/ccip/commit_reporting_plugin.go (hunk 5 above)
Comment (matYang) [reply core/services/ocr2/plugins/ccip/commit_reporting_plugin.go:812]: The `CommitStore` and `PriceRegistry` are ready to accept multi-gas price updates, gas prices being limited to 1 is an implementation detail of the current `commit_reporting_plugin`, from this perspective I would propose to confine the knowledge of `len(report.GasPrices) == 1` to inside `commit_reporting_plugin`

File: core/services/ocr2/plugins/ccip/internal/ccipdata/commit_store_v1_2_0.go (hunk 7 above)
Comment (matYang) [reply core/services/ocr2/plugins/ccip/internal/ccipdata/commit_store_v1_2_0.go:30]: Reusable functions are `IsBlessed`, `GetExpectedNextSequenceNumber`, `GetLatestPriceEpochAndRound`, `IsDown`, `VerifyExecutionReport`. These are simple functions, on the other hand embedding *CommitStoreV1_0_0 would require us to be careful with log poller registrations, it seems slightly hacky, overall I think it's worth it to duplicate a bit code here for simplicity.

File: contracts/src/v0.8/ccip/libraries/Internal.sol (hunk 3 above)
Comment (RensR) [reply contracts/src/v0.8/ccip/libraries/Internal.sol]: Yeah but the comment implies it does fit, so that needs to be changed

**Review by RensR — APPROVED** [review]
All good onchain

**Pull request merged**